src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
+            {
+                try
+                {
+                    RuntimeHelpers.EnsureSufficientExecutionStack();
+                }
+                catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public void Exit()
+        {
+            _recursionDepth--;
+        }
+
+        public void RunOnEmptyStack(Action<object> action, object state)
+        {
+            _executionStackCount++;
+
+            var recursionDepth = _recursionDepth;
+            _recursionDepth = 0;
+
+            try
+            {
+                // Using default scheduler rather than picking up the current scheduler.
+                var task = Task.Factory.StartNew(action, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+
+                // Task.Wait has the potential of inlining the task's execution on the current thread; avoid this.
+                ((IAsyncResult)task).AsyncWaitHandle.WaitOne();
```

Comment:
I'd expect it to take a while, though `VariableBinder` (on of the four places where this has to be applied) only does a read-only analysis of the tree, so there's a slim chance that one may complete quickly. I could move `GetAwaiter` up (we need it anyway), so the tiny bit of work involved in allocating the struct could overlap with the queued work, and use `IsCompleted` to check. Given that `IsCompleted` merely checks a flag, I don't see much harm in adding it other than a small increase in code size which I'm already offsetting by eliminating the two copies for `Task` and `Task<R>` (see previous comment).